avail-bootstrap/src/main/kotlin/com/avail/tools/bootstrap/SpecialObjectNamesGenerator.kt

```diff
@@ -0,0 +1,176 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of the contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.avail.tools.bootstrap
+
+import com.avail.AvailRuntime.Companion.specialObjects
+import com.avail.descriptor.representation.A_BasicObject
+import com.avail.descriptor.representation.NilDescriptor
+import com.avail.tools.bootstrap.Resources.escape
+import com.avail.tools.bootstrap.Resources.specialObjectCommentKey
+import com.avail.tools.bootstrap.Resources.specialObjectKey
+import com.avail.tools.bootstrap.Resources.specialObjectTypeKey
+import com.avail.tools.bootstrap.Resources.specialObjectsBaseName
+import java.io.PrintWriter
+import java.text.MessageFormat
+import java.util.*
+
+/**
+ * Generate a [property resource bundle][PropertyResourceBundle] that specifies
+ * unbound properties for the Avail names of the special objects.
+ *
+ * @author Todd L Smith &lt;todd@availlang.org&gt;
+ *
+ * @constructor
+ * Construct a new [SpecialObjectNamesGenerator].
+ *
+ * @param locale
+ *   The target [locale][Locale].
+ */
+internal class SpecialObjectNamesGenerator constructor(locale: Locale)
+	: PropertiesFileGenerator(specialObjectsBaseName, locale)
+{
+	/**
+	 * Write the names of the properties, whose unspecified values should be
+	 * the Avail names of the corresponding special objects.
+	 *
+	 * @param properties
+	 *   The existing [properties][Properties]. These should be copied into the
+	 *   resultant [properties resource bundle][ResourceBundle].
+	 * @param writer
+	 *   The [output stream][PrintWriter].
+	 */
+	override fun generateProperties(
+		properties: Properties,
+		writer: PrintWriter)
+	{
+		val specialObjects = specialObjects()
+		val keys = mutableSetOf<String>()
+		for (i in specialObjects.indices)
+		{
+			if (specialObjects[i] !== NilDescriptor.nil)
```

Comment:
While this happens to always work for the nil object in the current VM, this is performing an identity compare (like == in Java).  We don't generally do this with AvailObjects, preferring to use equals(), which is aliased to == in Kotlin.  That's because even objects with identity can have indirection objects that refer to them (and should be considered equal).  For nil, the preference is to use equalsNil().  There are a few more occurrences of this in this file.

Reply:
equalsNil() actually does the identity compare for its implementation:
`override fun equalsNil() = this === nil`
So how is using `!equalsNil()` different from `!==`
This is basically the reason I chose `!==`